FWCore/Framework/interface/CallbackBase.h

```diff
@@ -0,0 +1,270 @@
+#define FWCore_Framework_CallbackBase_h
+//
+// Package:     Framework
+// Class  :     CallbackBase
+//
+/**\class edm::eventsetup::CallbackBase
+
+ Description: Functional object used as the 'callback' for the CallbackProxy
+
+ Usage: Produces data objects for ESProducers in EventSetup system
+
+*/
+//
+// Author:      Chris Jones (original author, this was part of Callback.h),
+//              W. David Dagenhart (Refactored version + CallbackExternalWork, 2023)
+// Created:     Sun Apr 17 14:30:24 EDT 2005
+//
+
+#include <array>
+#include <atomic>
+#include <exception>
+#include <memory>
+#include <optional>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "oneapi/tbb/task_group.h"
+
+#include "FWCore/Concurrency/interface/SerialTaskQueueChain.h"
+#include "FWCore/Concurrency/interface/WaitingTask.h"
+#include "FWCore/Concurrency/interface/WaitingTaskHolder.h"
+#include "FWCore/Concurrency/interface/WaitingTaskList.h"
+#include "FWCore/Framework/interface/EventSetupImpl.h"
+#include "FWCore/Framework/interface/EventSetupRecordImpl.h"
+#include "FWCore/Framework/interface/produce_helpers.h"
+#include "FWCore/ServiceRegistry/interface/ActivityRegistry.h"
+#include "FWCore/ServiceRegistry/interface/ESModuleCallingContext.h"
+#include "FWCore/ServiceRegistry/interface/ESParentContext.h"
+#include "FWCore/ServiceRegistry/interface/ServiceRegistry.h"
+#include "FWCore/ServiceRegistry/interface/ServiceToken.h"
+#include "FWCore/Utilities/interface/ConvertException.h"
+#include "FWCore/Utilities/interface/ESIndices.h"
+#include "FWCore/Utilities/interface/Exception.h"
+#include "FWCore/Utilities/interface/Likely.h"
+#include "FWCore/Utilities/interface/propagate_const.h"
+#include "FWCore/Utilities/interface/Signal.h"
+
+namespace edm {
+  void exceptionContext(cms::Exception&, ESModuleCallingContext const&);
+
+  namespace eventsetup {
+
+    // The default decorator that does nothing
+    template <typename TRecord>
+    struct CallbackSimpleDecorator {
+      void pre(const TRecord&) {}
+      void post(const TRecord&) {}
+    };
+
+    template <typename T,             //producer's type
+              typename TProduceFunc,  //produce functor type
+              typename TReturn,       //return type of the producer's method
+              typename TRecord,       //the record passed in as an argument
+              typename TDecorator>    //allows customization using pre/post calls
+    class CallbackBase {
+    public:
+      CallbackBase(T* iProd, std::shared_ptr<TProduceFunc> iProduceFunc, unsigned int iID, const TDecorator& iDec)
+          : proxyData_{},
+            producer_(iProd),
+            callingContext_(&iProd->description()),
+            produceFunction_(std::move(iProduceFunc)),
+            id_(iID),
+            wasCalledForThisRecord_(false),
+            decorator_(iDec) {}
+
+      CallbackBase(const CallbackBase&) = delete;
+      const CallbackBase& operator=(const CallbackBase&) = delete;
```

Comment:
Hmmmm.

You are certainly correct about the return type. I changed it and will push it into GitHub soon (I'll push all 4 commits at the same time when they are done). I had copied/pasted this from Callback.h and just changed the name without even thinking... Note that this error occurs often in FWCore/Framework. I didn't test but it might even be in the FWCore/Skeletons code generation. It is rare for this to make any practical difference, but it is incorrect.

About the move operators. I am leaning towards leaving them out. Deleting the copy constructor/assignment also deletes the move operators unless you explicitly define them so it seems like unnecessary boilerplate to me. In FWCore/Framework the vast majority of the time that is the pre-existing pattern. The rule of five in the CMS Style Guide says  "definition of any of the following requires definition of all five", but deletion is not definition... I don't feel strongly about this.

Reply:
I interpret the rule of five such that deletion is one kind of definition. I find the rules for compiler-generated copy/move operators complicated, and prefer to define all of them explicitly (if any of them need to be defined) so that the intended behavior is clear. Although in practice the `std::atomic<bool>` member effectively disables all copy and move operations anyway.

Reply:
Done. I followed your interpretation of the rule of 5. I also cleaned up the other classes I significantly modified in this PR related to this and the assignment return value. (There are lots of other files in the Framework that have the same issues.)

Reply:
Thanks!

> (There are lots of other files in the Framework that have the same issues.)

I'd bet many of them predate C++11 and the move construction/assignment.